src/libstore/store-api.cc

```diff
@@ -301,11 +309,11 @@ void Store::queryPathInfo(const StorePath & storePath,
 
         {
             auto res = state.lock()->pathInfoCache.get(hashPart);
-            if (res) {
+            if (res && res->isKnownNow()) {
                 stats.narInfoReadAverted++;
-                if (!*res)
+                if (!res->didExist())
                     throw InvalidPath("path '%s' is not valid", printStorePath(storePath));
-                return callback(ref<const ValidPathInfo>(*res));
+                callback(ref<const ValidPathInfo>(res->value));
```

Comment:
Hm, removing the return looks potentially wrong...

Reply:
Yes it was. The compiler was complaining about `return` in a `void` function. I've re-added the return and then it works.